app/multitenant/sqs_control_router.go

```diff
@@ -0,0 +1,330 @@
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"golang.org/x/net/context"
+
+	"github.com/weaveworks/scope/app"
+	"github.com/weaveworks/scope/common/xfer"
+)
+
+var (
+	longPollTime = aws.Int64(10)
+	rpcTimeout   = time.Minute
+)
+
+// sqsControlRouter:
+// Creates a queue for every probe that connects to it, and a queue for
+// responses back to it.  When it recieves a request, posts it to the
+// probe queue.  When probe recieves a request, handles it and posts the
+// response back to the response queue.
+type sqsControlRouter struct {
+	service  *sqs.SQS
+	queueURL *string
+	userIDer UserIDer
+
+	mtx          sync.Mutex
+	responses    map[string]chan xfer.Response
+	probeWorkers map[int64]*probeWorker
+}
+
+type sqsRequestMessage struct {
+	ID               string
+	Request          xfer.Request
+	ResponseQueueURL string
+}
+
+type sqsResponseMessage struct {
+	ID       string
+	Response xfer.Response
+}
+
+// NewSQSControlRouter the harbinger of death
+func NewSQSControlRouter(url, region string, creds *credentials.Credentials, userIDer UserIDer) app.ControlRouter {
+	result := &sqsControlRouter{
+		service: sqs.New(session.New(aws.NewConfig().
+			WithEndpoint(url).
+			WithRegion(region).
+			WithCredentials(creds))),
+		queueURL:     nil,
+		userIDer:     userIDer,
+		responses:    map[string]chan xfer.Response{},
+		probeWorkers: map[int64]*probeWorker{},
+	}
+	go result.loop()
+	return result
+}
+
+func (cr *sqsControlRouter) Stop() error {
+	return nil
+}
+
+func (cr *sqsControlRouter) setQueueURL(url *string) {
+	cr.mtx.Lock()
+	defer cr.mtx.Unlock()
+	cr.queueURL = url
+}
+
+func (cr *sqsControlRouter) getQueueURL() *string {
+	cr.mtx.Lock()
+	defer cr.mtx.Unlock()
+	return cr.queueURL
+}
+
+func (cr *sqsControlRouter) getOrCreateQueue(name string) (*string, error) {
+	getQueueURLRes, err := cr.service.GetQueueUrl(&sqs.GetQueueUrlInput{
+		QueueName: aws.String(name),
+	})
+	if err == nil {
+		return getQueueURLRes.QueueUrl, nil
+	}
+
+	createQueueRes, err := cr.service.CreateQueue(&sqs.CreateQueueInput{
+		QueueName: aws.String(name),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return createQueueRes.QueueUrl, nil
+}
+
+func (cr *sqsControlRouter) loop() {
+	for {
+		// This app has a random id and uses this as a return path for all responses from probes.
+		name := fmt.Sprintf("control-app-%d", rand.Int63())
+		queueURL, err := cr.getOrCreateQueue(name)
+		if err != nil {
+			log.Errorf("Failed to create queue: %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		cr.setQueueURL(queueURL)
+		break
+	}
+
+	for {
+		res, err := cr.service.ReceiveMessage(&sqs.ReceiveMessageInput{
+			QueueUrl:        cr.queueURL,
+			WaitTimeSeconds: longPollTime,
+		})
+		if err != nil {
+			log.Errorf("Error recieving message from %s: %v", *cr.queueURL, err)
+			continue
+		}
+		if len(res.Messages) == 0 {
+			continue
+		}
+		cr.handleResponses(res)
+		if err := cr.deleteMessages(cr.queueURL, res.Messages); err != nil {
```

Comment:
Ah yeah, but deleteMessages will delete from any queue, while cr.queueURL is the url of the queue the app is listening on for ui response (as oppose to probe requests).  So it needs to be passed about.